torch/nn/modules/module.py

```diff
@@ -1494,10 +1526,10 @@ def _slow_forward(self, *input, **kwargs):
     def _call_impl(self, *args, **kwargs):
         forward_call = (self._slow_forward if torch._C._get_tracing_state() else self.forward)
         # If we don't have any hooks, we want to skip the rest of the logic in
-        # this function, and just call forward.
-        if not (self._backward_hooks or self._backward_pre_hooks or self._forward_hooks or self._forward_pre_hooks
-                or _global_backward_pre_hooks or _global_backward_hooks
-                or _global_forward_hooks or _global_forward_pre_hooks):
+        # this function, and just call forward.  It's slow for dynamo to guard on the state
+        # of all these hook dicts individually, so instead it can guard on 2 bools and we just
+        # have to promise to keep them up to date when hooks are added or removed via official means.
+        if not self._has_hooks and not _has_global_hooks:
```

Comment:
It's not ok to assume that `_has_hooks` is available as some people pickle their Module directly. You need to update the `__setstate__`  function to properly populate this field if it is not in the serialized state.

Reply:
ok, good point. i believe i should add a test case for pickle/unpickle and ensure that __setstate__ correctly enforces that _update_has_hooks gets installed and run.

would that address your concern?

Reply:
fixing setstate will work yes.
Unfortunately testing this is challenging indeed :/

Reply:
Fix via: https://github.com/pytorch/pytorch/pull/96224